File: src/generic/incremental_registration.cpp
#include <pcl/point_cloud.h>
#include <pcl/registration/incremental_registration.h>

#define IncrementalReg_ptr boost::shared_ptr<pcl::registration::IncrementalRegistration<_PointT, float> >
#define ICP_ptr boost::shared_ptr<pcl::IterativeClosestPoint<_PointT, _PointT> >
#define ICPNL_ptr boost::shared_ptr<pcl::IterativeClosestPointNonLinear<_PointT, _PointT> >
#define PointCloud_ptr pcl::PointCloud<_PointT>::Ptr

PCLIMP(void*, IncrementalRegistration, new)()
{
  return new IncrementalReg_ptr(new pcl::registration::IncrementalRegistration<_PointT, float>());
}

PCLIMP(void, IncrementalRegistration, delete)(IncrementalReg_ptr *self)
{
  delete self;
}

PCLIMP(void, IncrementalRegistration, setICP)(IncrementalReg_ptr *self, ICP_ptr *icp)
{
  (*self)->setRegistration(*icp);
}

PCLIMP(void, IncrementalRegistration, setICPNL)(IncrementalReg_ptr *self, ICPNL_ptr *icpnl)
{
  (*self)->setRegistration(*icpnl);
}

PCLIMP(void, IncrementalRegistration, reset)(IncrementalReg_ptr *self)
{
  (*self)->reset();
}

PCLIMP(bool, IncrementalRegistration, registerCloud)(IncrementalReg_ptr *self, PointCloud_ptr *cloud, THFloatTensor* delta_estimate)
{
  Eigen::Matrix4f _delta_estimate = Eigen::Matrix<float, 4, 4>::Identity();
  if (delta_estimate)
    _delta_estimate = Tensor2Mat<4,4>(delta_estimate);
  return (*self)->registerCloud(*cloud, _delta_estimate);
}

PCLIMP(void, IncrementalRegistration, getDeltaTransform)(IncrementalReg_ptr *self, THFloatTensor* output)
{
  Eigen::Matrix4f m = (*self)->getDeltaTransform();
  copyMatrix(m, output);
}

PCLIMP(void, IncrementalRegistration, getAbsoluteTransform)(IncrementalReg_ptr *self, THFloatTensor* output)
{
  Eigen::Matrix4f m = (*self)->getAbsoluteTransform();
  copyMatrix(m, output);
}

#undef IncrementalReg_ptr
#undef ICP_ptr
#undef ICPNL_ptr
#undef PointCloud_ptr

File: IncrementalRegistration.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local IncrementalRegistration = torch.class('pcl.IncrementalRegistration', pcl)

local func_by_type = {}

local function init()

  local IncrementalRegistration_method_names = {
    'new',
    'delete',
    'setICP',
    'setICPNL',
    'registerCloud',
    'getDeltaTransform',
    'getAbsoluteTransform'
  }

  for k,v in pairs(utils.type_key_map) do
    func_by_type[k] = utils.create_typed_methods("pcl_IncrementalRegistration_TYPE_KEY_", IncrementalRegistration_method_names, v)
  end
end

init()

function IncrementalRegistration:__init(pointType)
  pointType = pointType or pcl.PointXYZ
  self.pointType = pcl.pointType(pointType)
  self.f = func_by_type[self.pointType]
  self.o = self.f.new()
end

function IncrementalRegistration:cdata()
  return self.o
end

function IncrementalRegistration:setRegistration(reg)
  if torch.isTypeOf(reg, pcl.ICP) then
    self.f.setICP(self.o, reg:cdata())
  elseif torch.isTypeOf(reg, pcl.ICPNL) then
    self.f.setICPNL(self.o, reg:cdata())
  else
    utils.check_arg('reg', false, 'Unsupported registration algorithm obj')
  end
end

function IncrementalRegistration:reset()
  self.f.reset(self.o)
end

function IncrementalRegistration:registerCloud(cloud, delta_estimate)
  return self.f.registerCloud(self.o, cloud:cdata(), delta_estimate)
end

function IncrementalRegistration:getDeltaTransform()
  local t = torch.FloatTensor()
  self.f.getDeltaTransform(self.o, t:cdata())
  return t
end

function IncrementalRegistration:getAbsoluteTransform()
  local t = torch.FloatTensor()
  self.f.getAbsoluteTransform(self.o, t:cdata())
  return t
end
